fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/jbatch/bulkdata/export/fast/ResourcePayloadReader.java

```diff
@@ -0,0 +1,683 @@
+import com.ibm.cloud.objectstorage.services.s3.model.CreateBucketRequest;
+import com.ibm.cloud.objectstorage.services.s3.model.PartETag;
+import com.ibm.fhir.config.FHIRConfigHelper;
+import com.ibm.fhir.config.FHIRConfiguration;
+import com.ibm.fhir.config.FHIRRequestContext;
+import com.ibm.fhir.jbatch.bulkdata.common.BulkDataUtils;
+import com.ibm.fhir.jbatch.bulkdata.common.Constants;
+import com.ibm.fhir.model.resource.Resource;
+import com.ibm.fhir.model.util.ModelSupport;
+import com.ibm.fhir.persistence.FHIRPersistence;
+import com.ibm.fhir.persistence.ResourcePayload;
+import com.ibm.fhir.persistence.helper.FHIRPersistenceHelper;
+import com.ibm.fhir.persistence.helper.FHIRTransactionHelper;
+import com.ibm.fhir.search.date.DateTimeHandler;
+
+/**
+ * A high-performance version of the bulk-export job which doesn't support typeFilter and so
+ * can use a more efficient mechanism to page through the set of resources filtered by _lastModified.
+ *
+ * The Java Batch framework separates the reader from the writer. On the face of it, separation of
+ * concerns seems like a good idea. However, because the fetching of the data is controlled by the
+ * persistence layer and the fact that we may need to end one multi-part upload and continue feeding
+ * a new multi-part upload, the separation makes things a lot harder and less readable - there is a
+ * lack of cohesion.
+ *
+ * To address this, we handle both reading and writing in this class, leaving the ItemWriter as a
+ * minimal stub. The framework is still leveraged for checkpointing.
+ *
+ * The {@link #readItem()} call simply returns an Object if there is more data to process. The
+ * {@link ResourceExportCheckpointAlgorithm} will always say we need a checkpoint, so the
+ * checkpoint will occur immediately after each read/write.
+ *
+ * The export is complete when {@link #readItem()} returns null.
+ *
+ * The export assumes that the persistence layer can efficiently scan forward based on the last_updated
+ * time of a resource. It is possible that multiple resources may share the same last_updated time, so
+ * this class has to track which resources have already been processed for a given timestamp. This is
+ * fairly easy, because the persistence layer must provide the data in order of last_updated. Because
+ * the number of resources for a given timestamp is probably very small, it is more efficient to track
+ * it this way than asking the persistence layer (e.g. JDBC) to sort. This can cause performance issues,
+ * because it negates the benefit of traversing an index in order.
+ */
+@Dependent
+public class ResourcePayloadReader extends AbstractItemReader {
+    private final static Logger logger = Logger.getLogger(ResourcePayloadReader.class.getName());
+    private static final String CLASS = ResourcePayloadReader.class.getName();
+
+    // S3 client API to IBM Cloud Object Storage
+    private AmazonS3 cosClient = null;
+
+    // TODO ???
+    boolean isSingleCosObject = false;
```

Comment:
We can probably just delete this from here since its not used.  At one point, I think our other impl has a setting for exporting 1-object-per-type vs splitting a single type across multiple objects, but I don't know if thats really needed...it seems if you really want all the resources of a given type in one huge file you could just set the limits to very large values.

Reply:
+1